bothan-api/server-cli/src/commands/query.rs

```diff
@@ -0,0 +1,433 @@
+use std::collections::{HashMap, HashSet};
+use std::error::Error as StdError;
+use std::time::Duration;
+
+use anyhow::{Context, anyhow};
+use bothan_api::config::AppConfig;
+use bothan_lib::types::AssetInfo;
+use bothan_lib::worker::rest::AssetInfoProvider as RestAssetInfoProvider;
+use bothan_lib::worker::websocket::{
+    AssetInfoProvider as WebSocketAssetInfoProvider, AssetInfoProviderConnector, Data,
+};
+use clap::{Args, Parser, Subcommand};
+use futures::stream::{FuturesOrdered, StreamExt};
+use humantime::Duration as HumanDuration;
+use itertools::Itertools;
+use prettytable::{Table, row};
+use tokio::time::timeout;
+
+const DEFAULT_TIMEOUT: &str = "10s";
+const CONFIG_ERROR_MESSAGE: &str = "config not found in config.toml, please refer to the example at /bothan-api/server-cli/config.example.toml";
+
+#[derive(Parser)]
+pub struct QueryCli {
+    #[command(subcommand)]
+    subcommand: QuerySubCommand,
+}
+
+#[derive(Args, Debug, Clone)]
+pub struct QueryArgs {
+    /// The list of query ids to query prices for
+    pub query_ids: Vec<String>,
+
+    /// Timeout duration
+    #[arg(short, long, default_value = DEFAULT_TIMEOUT)]
+    pub timeout: HumanDuration,
+}
+
+#[derive(Subcommand, Debug)]
+pub enum QuerySubCommand {
+    /// Query Binance prices
+    Binance {
+        #[clap(flatten)]
+        args: QueryArgs,
+    },
+    /// Query Bitfinex prices
+    Bitfinex {
+        #[clap(flatten)]
+        args: QueryArgs,
+    },
+    /// Query Bybit prices
+    Bybit {
+        #[clap(flatten)]
+        args: QueryArgs,
+    },
+    /// Query Coinbase prices
+    Coinbase {
+        #[clap(flatten)]
+        args: QueryArgs,
+    },
+    /// Query Coingecko prices
+    Coingecko {
+        #[clap(flatten)]
+        args: QueryArgs,
+    },
+    /// Query Coinmarketcap prices
+    Coinmarketcap {
+        #[clap(flatten)]
+        args: QueryArgs,
+    },
+    /// Query HTX prices
+    Htx {
+        #[clap(flatten)]
+        args: QueryArgs,
+    },
+    /// Query Kraken prices
+    Kraken {
+        #[clap(flatten)]
+        args: QueryArgs,
+    },
+    /// Query OKX prices
+    Okx {
+        #[clap(flatten)]
+        args: QueryArgs,
+    },
+}
+
+impl QueryCli {
+    pub async fn run(&self, app_config: AppConfig) -> anyhow::Result<()> {
+        match &self.subcommand {
+            QuerySubCommand::Binance { args } => {
+                let opts = app_config
+                    .manager
+                    .crypto
+                    .source
+                    .binance
+                    .with_context(|| format!("Binance {}", CONFIG_ERROR_MESSAGE))?;
```

Comment:
i dont think we need to wrap this with context, we can just return an explicit config is missing error